src/librustc_mir/transform/fragment_locals.rs

```diff
@@ -0,0 +1,390 @@
+    ) -> (&'a mut Self, &'tcx [PlaceElem<'tcx>]) {
+        let mut variant_index = None;
+        while let [elem, rest @ ..] = proj_elems {
+            if let FragmentTreeNodeKind::OpaqueLeaf { .. } = self.kind {
+                break;
+            }
+
+            match *elem {
+                ProjectionElem::Field(f, ty) => {
+                    let field = Fragment::Field(variant_index, f);
+                    // FIXME(eddyb) use `self.fragment(field)` post-Polonius(?).
+                    match self.kind {
+                        FragmentTreeNodeKind::Nested(ref mut fragments) => {
+                            self = fragments.entry(field).or_insert_with(|| FragmentTree::new(ty));
+                        }
+                        FragmentTreeNodeKind::OpaqueLeaf { .. } => unreachable!(),
+                    }
+                }
+
+                ProjectionElem::Downcast(..) => {}
+
+                // FIXME(eddyb) support indexing by constants.
+                ProjectionElem::ConstantIndex { .. } | ProjectionElem::Subslice { .. } |
+                // Can't support without alias analysis.
+                ProjectionElem::Index(_) | ProjectionElem::Deref => {
+                    // If we can't project, we must be opaque.
+                    self.make_opaque();
+                    break;
+                }
+            }
+
+            proj_elems = rest;
+            variant_index = match *elem {
+                ProjectionElem::Downcast(_, v) => Some(v),
+                _ => None,
+            };
+        }
+
+        (self, proj_elems)
+    }
+
+    fn assign_locals(
+        &mut self,
+        local_decls: &mut IndexVec<Local, LocalDecl<'tcx>>,
+        source_info: SourceInfo,
+    ) {
+        match self.kind {
+            FragmentTreeNodeKind::OpaqueLeaf { ref mut replacement_local } => {
+                let mut decl = LocalDecl::new_internal(self.ty, source_info.span);
+                decl.source_info = source_info;
+                *replacement_local = Some(local_decls.push(decl));
+            }
+            FragmentTreeNodeKind::Nested(ref mut fragments) => {
+                for fragment in fragments.values_mut() {
+                    fragment.assign_locals(local_decls, source_info);
+                }
+            }
+        }
+    }
+}
+
+struct FragmentTreeCollector<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    locals: IndexVec<Local, FragmentTree<'tcx>>,
+}
+
+impl FragmentTreeCollector<'tcx> {
+    fn place_node(&'a mut self, place: &Place<'tcx>) -> Option<&'a mut FragmentTree<'tcx>> {
+        let (node, proj_elems) = self.locals[place.local].project(place.projection);
+        if proj_elems.is_empty() { Some(node) } else { None }
+    }
+}
+
+impl Visitor<'tcx> for FragmentTreeCollector<'tcx> {
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, _: Location) {
+        if let Some(node) = self.place_node(place) {
+            if context.is_use() {
+                node.make_opaque();
```

Comment:
Note to self: if the context is a borrow, this should make the whole local "opaque" (i.e. indivisible - maybe I should use that word in the implementation?).

cc @RalfJung (I think this is the conservative thing to do, as it would keep letting unsafe code access sibling fields, of those borrowed - whereas the current state of the PR would make some examples UB, because the borrowed field may end up a separate local)

Reply:
Sorry, I don't have nearly enough context to comprehend what you wrote. My best guess is "if the address of the local is never taken then we don't do the splitting", and that seems suitably conservative indeed. :D